iavl/iavl_range_proofs.go

```diff
@@ -0,0 +1,400 @@
+	if proof.PathToKey.isLeftmost() {
+		return nil
+	}
+	// The start key is in between the left path and the key returned,
+	// and the paths are adjacent. Therefore there is nothing between
+	// the key returned and the start key.
+	if proof.Left != nil && proof.Left.Path.isLeftAdjacentTo(proof.PathToKey) {
+		return nil
+	}
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyLastInRangeProof is a proof that a given key is the last in a given range.
+type KeyLastInRangeProof struct {
+	KeyExistsProof `json:"key_proof"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// String returns a string representation of the proof.
+func (proof *KeyLastInRangeProof) String() string {
+	// TODO(cloudhead): Needs work.
+	return fmt.Sprintf("%#v", proof)
+}
+
+// Verify that the last in range proof is valid.
+func (proof *KeyLastInRangeProof) Verify(startKey, endKey, key, value []byte, root []byte) error {
+	if key != nil && (bytes.Compare(key, startKey) == -1 || bytes.Compare(key, endKey) == 1) {
+		return ErrInvalidInputs
+	}
+	if proof.Left == nil && proof.Right == nil && proof.PathToKey == nil {
+		return ErrInvalidProof
+	}
+	if err := verifyPaths(proof.Left, proof.Right, startKey, endKey, root); err != nil {
+		return err
+	}
+	if proof.PathToKey == nil {
+		// If we don't have an existing key, we effectively have a proof of absence.
+		return verifyKeyAbsence(proof.Left, proof.Right)
+	}
+
+	if err := proof.KeyExistsProof.Verify(key, value, root); err != nil {
+		return err
+	}
+	if bytes.Equal(key, endKey) {
+		return nil
+	}
+	if proof.PathToKey.isRightmost() {
+		return nil
+	}
+	if proof.Right != nil &&
+		proof.PathToKey.isLeftAdjacentTo(proof.Right.Path) {
+		return nil
+	}
+
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// Verify that a range proof is valid.
+//
+// This method expects the same parameters passed to query the range.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return ErrInvalidInputs
+	}
+	if limit > 0 && len(keys) > limit {
+		return ErrInvalidInputs
+	}
+	if len(proof.PathToKeys) == 0 && proof.Left == nil && proof.Right == nil {
+		return ErrInvalidProof
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	startKey, endKey, keys, values = reverseIfDescending(startKey, endKey, keys, values)
+
+	{
+		firstKey, lastKey := startKey, endKey
+		if limit > 0 && len(keys) == limit {
```

Comment:
i thought you modified this logic here as well with the extra test case for the partial proof issue.
or did you solve this some other way?